micro-rdk/src/common/conn/viam.rs

```diff
@@ -734,35 +774,58 @@ where
+                let mut app_client_tasks: FuturesUnordered<_> = FuturesUnordered::new();
                 for task in &self.app_client_tasks {
-                    app_client_tasks.push(AppClientTaskRunner {
-                        app_client: &app_client,
-                        invoker: task,
-                        state: TaskRunnerState::Run {
-                            task: task.invoke(&app_client),
-                        },
+                    app_client_tasks.push({
+                        async move {
+                            let mut duration = task.get_default_period();
+                            loop {
+                                let _ = Timer::after(duration).await;
+                                match task.invoke(app_client).await {
+                                    Ok(Some(next_duration)) => {
+                                        duration = next_duration;
+                                    }
+                                    Ok(None) => {}
+                                    e @ Err(_) => {
+                                        break e;
+                                    }
+                                }
+                            }
+                        }
                     });
                 }
                 while let Some(res) = app_client_tasks.next().await {
                     if let Err(err) = res {
-                        log::error!("a task returned the following error {:?}", err);
+                        log::error!(
+                            "an app client task returned an error {:?} - dropping app client",
+                            err
+                        );
                         break;
                     }
                 }
             }
+
+            // Explicitly release the app_client resources before sleeping, and before
+            // trying to construct a new one with `connect_to_app` below.
+            std::mem::drop(app_client.take());
+            let _ = Timer::after(wait).await;
+
             // the only way to reach here is either we had a None passed (app_client wasn't connected at boot)
             // or an error was reported by an underlying task which means that app client
             // is considered gone
-            let _ = Timer::after(wait).await;
+
+            log::info!("attempting to reestablish a connection to app for app client tasks");
             app_client = self
                 .connect_to_app()
                 .await
                 .inspect_err(|error| {
-                    if error.is_permission_denied() || error.is_unauthenticated() {
+                    log::error!("couldn't connect to app server, reason {:?}", error);
+                     if error.is_permission_denied() || error.is_unauthenticated() {
+                        log::error!("credentials were positively rejected by app - cached credentials and configuration will be erased");
```

Comment:
same as comment above on similar line. these checks around here and above where I commented might be a candidate for DRY